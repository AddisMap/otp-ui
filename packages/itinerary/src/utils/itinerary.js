export const transitModes = [
  "TRAM",
  "BUS",
  "SUBWAY",
  "FERRY",
  "RAIL",
  "GONDOLA"
];

export function isTransit(mode) {
  return transitModes.includes(mode) || mode === "TRANSIT";
}

// TODO: Is this a place where we can get these colors dynamically?
export function getMapColor(mode) {
  switch (mode) {
    case "WALK":
      return "#444";
    case "BICYCLE":
      return "#0073e5";
    case "SUBWAY":
      return "#f00";
    case "RAIL":
      return "#b00";
    case "BUS":
      return "#080";
    case "TRAM":
      return "#800";
    case "FERRY":
      return "#008";
    case "CAR":
      return "#444";
    case "MICROMOBILITY":
      return "#f5a729";
    default:
      return "#aaa";
  }
}

// TODO: temporary code; handle via migrated OTP i18n language table
export function getStepDirection(step) {
  switch (step.relativeDirection) {
    case "DEPART":
      return `Head ${step.absoluteDirection.toLowerCase()}`;
    case "LEFT":
      return "Left";
    case "HARD_LEFT":
      return "Hard left";
    case "SLIGHTLY_LEFT":
      return "Slight left";
    case "CONTINUE":
      return "Continue";
    case "SLIGHTLY_RIGHT":
      return "Slight right";
    case "RIGHT":
      return "Right";
    case "HARD_RIGHT":
      return "Hard right";
    case "CIRCLE_CLOCKWISE":
      return "Follow circle clockwise";
    case "CIRCLE_COUNTERCLOCKWISE":
      return "Follow circle counterclockwise";
    case "ELEVATOR":
      return "Take elevator";
    case "UTURN_LEFT":
      return "Left U-turn";
    case "UTURN_RIGHT":
      return "Right U-turn";
    default:
      return step.relativeDirection;
  }
}

export function getStepInstructions(step) {
  const conjunction = step.relativeDirection === "ELEVATOR" ? "to" : "on";
  return `${getStepDirection(step)} ${conjunction} ${step.streetName}`;
}

export function getStepStreetName(step) {
  if (step.streetName === "road") return "Unnamed Road";
  if (step.streetName === "path") return "Unnamed Path";
  return step.streetName;
}

/**
 * Get the configured company object for the given network string if the company
 * has been defined in the provided companies array config.
 */
function getCompanyForNetwork(networkString, companies = []) {
  const company = companies.find(co => co.id === networkString);
  if (!company) {
    console.warn(
      `No company found in config.yml that matches rented vehicle network: ${networkString}`,
      companies
    );
  }
  return company;
}

/**
 * Returns mode name by checking the vertex type (VertexType class in OTP) for
 * the provided place. NOTE: this is currently only intended for vehicles at
 * the moment (not transit or walking).
 *
 * TODO: I18N
 * @param  {string} place place from itinerary leg
 */
export function getModeForPlace(place) {
  switch (place.vertexType) {
    case "CARSHARE":
      return "car";
    case "VEHICLERENTAL":
      return "E-scooter";
    // TODO: Should the type change depending on bike vertex type?
    case "BIKESHARE":
    case "BIKEPARK":
      return "bike";
    // If company offers more than one mode, default to `vehicle` string.
    default:
      return "vehicle";
  }
}

export function getPlaceName(place, companies) {
  // If address is provided (i.e. for carshare station, use it)
  if (place.address) return place.address.split(",")[0];
  if (place.networks && place.vertexType === "VEHICLERENTAL") {
    // For vehicle rental pick up, do not use the place name. Rather, use
    // company name + vehicle type (e.g., SPIN E-scooter). Place name is often just
    // a UUID that has no relevance to the actual vehicle. For bikeshare, however,
    // there are often hubs or bikes that have relevant names to the user.
    const company = getCompanyForNetwork(place.networks[0], companies);
    if (company) {
      return `${company.label} ${getModeForPlace(place)}`;
    }
  }
  // Default to place name
  return place.name;
}

export function toSentenceCase(str) {
  if (str == null) {
    return "";
  }
  const typedString = String(str);
  return (
    typedString.charAt(0).toUpperCase() + typedString.substr(1).toLowerCase()
  );
}

export function getLegModeLabel(leg) {
  switch (leg.mode) {
    case "BICYCLE_RENT":
      return "Biketown";
    case "CAR":
      return leg.hailedCar ? "Ride" : "Drive";
    case "GONDOLA":
      return "Aerial Tram";
    case "TRAM":
      if (leg.routeLongName.toLowerCase().indexOf("streetcar") !== -1)
        return "Streetcar";
      return "Light Rail";
    case "MICROMOBILITY":
      return "Ride";
    default:
      return toSentenceCase(leg.mode);
  }
}

/** Looks up an operator from the provided configuration */
export const getOperatorFromConfig = (id, config) =>
  config.operators.find(operator => operator.id === id) || null;
