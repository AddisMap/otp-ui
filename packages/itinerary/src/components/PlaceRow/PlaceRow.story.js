import React from "react";
import { storiesOf } from "@storybook/react";
import { withA11y } from "@storybook/addon-a11y";
import { withInfo } from "@storybook/addon-info";
import { withThemesProvider } from "storybook-addon-styled-component-theme";
import { action } from "@storybook/addon-actions";
import PlaceRow from ".";
import MainStyleDecorator from "../../common/components/MainStyleDecorator";
import {
  LIGHT_THEME,
  DARK_THEME
} from "../../common/constants";
import ThemeShim from "../../common/components/ThemeShim";
import itinerary from "../../__mocks__/itinerary";
import config from "../../__mocks__/config";

const accessLegIndex = 0;
const transitLegIndex = 1;
const accessLeg = itinerary.legs[accessLegIndex];
const transitLeg = itinerary.legs[transitLegIndex];
const customIcons = [];
// const place = itinerary.legs[0].from;
// const time = itinerary.legs[0].startTime;
const timeOptions = { format: "h:mm a" };
const followsTransit = false;
const routingType = "ITINERARY";

// Originally from OTP MOD config
const longDateFormat = "MMMM D, YYYY";
// Originally from OTP MOD config
const timeFormat = "h:mm a";

const setActiveLeg = action("setActiveLeg");
const frameLeg = action("frameLeg");
const toRouteAbbreviation = r => r.toString();

storiesOf("Itinerary|PlaceRow", module)
  // .addDecorator(withThemesProvider([LIGHT_THEME, DARK_THEME]))
  .addDecorator(withA11y)
  .addDecorator(withInfo)
  .addDecorator(MainStyleDecorator)
  .add("Place Row with Access Leg", () => (
    <ThemeShim>
      <PlaceRow
        leg={accessLeg}
        legIndex={accessLegIndex}
        place={accessLeg.from}
        time={accessLeg.startTime}
        config={config}
        customIcons={customIcons}
        timeOptions={timeOptions}
        followsTransit={followsTransit}
        routingType={routingType}
        setActiveLeg={setActiveLeg}
        frameLeg={frameLeg}
        toRouteAbbreviation={toRouteAbbreviation}
      />
    </ThemeShim>
  ))
  .add("Place Row with Transit Leg", () => (
    <ThemeShim>
      <PlaceRow
        leg={transitLeg}
        legIndex={transitLegIndex}
        place={transitLeg.from}
        time={transitLeg.startTime}
        config={config}
        customIcons={customIcons}
        timeOptions={timeOptions}
        followsTransit={followsTransit}
        routingType={routingType}
        setActiveLeg={setActiveLeg}
        frameLeg={frameLeg}
        longDateFormat={longDateFormat}
        timeFormat={timeFormat}
        toRouteAbbreviation={toRouteAbbreviation}
      />
    </ThemeShim>
  ));
