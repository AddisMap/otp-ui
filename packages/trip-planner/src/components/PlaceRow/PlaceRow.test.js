import React from "react";
import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import MOCK_THEME from "~/packages/trip-planner/src/common/__mocks__/theme";
import PlaceRow from ".";
import itinerary from "../../__mocks__/itinerary";
import config from "../../__mocks__/config";

describe("Place Row Component", () => {
  let sut;

  const customIcons = [];
  const leg = itinerary.legs[0];
  const legIndex = 0;
  const place = itinerary.legs[0].from;
  const time = itinerary.legs[0].startTime;
  const timeOptions = { format: "h:mm a" };
  const followsTransit = false;
  const routingType = "ITINERARY";
  const setActiveLeg = jest.fn(() => {});
  const frameLeg = jest.fn(() => {});
  const toRouteAbbreviation = jest.fn(r => r.toString());

  beforeEach(() => {
    sut = render(
      <ThemeProvider theme={MOCK_THEME}>
        <PlaceRow
          config={config}
          leg={leg}
          customIcons={customIcons}
          legIndex={legIndex}
          place={place}
          time={time}
          timeOptions={timeOptions}
          followsTransit={followsTransit}
          routingType={routingType}
          setActiveLeg={setActiveLeg}
          frameLeg={frameLeg}
          toRouteAbbreviation={toRouteAbbreviation}
        />
      </ThemeProvider>
    );
  });

  afterEach(() => jest.clearAllMocks());

  it("should match the snapshot", () => {
    const { asFragment } = sut;
    expect(asFragment()).toMatchSnapshot();
  });
});
