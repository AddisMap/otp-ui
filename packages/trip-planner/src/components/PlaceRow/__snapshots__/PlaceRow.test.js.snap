// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Place Row Component should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-htpNat dluyAv"
  >
    <div
      class="sc-bxivhb hjAAiM"
    >
      4:42 pm
    </div>
    <div
      class="sc-ifAKCX fLNOhX"
    >
      <div
        class="sc-htoDjs hIELFT"
      >
        <div
          class="sc-dnqmqq kfYVVq"
          mode="WALK"
        />
        <div
          class="sc-iwsKbI dZIZQa"
        >
          <div
            aria-label="Travel by WALK"
            class="sc-gZMcBi iObVEa"
            mode="WALK"
          >
            <svg
              class="sc-gqjmRU hnZXkc"
              fill="black"
              height="18"
              mode="WALK"
              viewBox="0 0 390 390"
              width="18"
            >
              <title>
                Travel by WALK
              </title>
              <path
                d="M209 0c17.8 0 32.3 14.5 32.3 32.3S226.8 64.6 209 64.6s-32.3-14.5-32.3-32.3S191.2 0 209 0zm98.5 201.8c-4.5 6.6-13.5 8.3-20.1 3.8 0 0-44.7-30.5-47.9-32.7-3.2-2.2-5.1-6.8-6.5-9.8-1-2-6-11.8-9.3-18.3l-4 17-10.4 47.1s55.7 65.6 56.8 66.9c1.6 1.8 2.5 5.5 3.2 8.3.7 2.7 14.9 79.9 14.9 79.9 2.2 11.9-5.7 23.4-17.6 25.7-11.9 2.2-23.4-5.7-25.7-17.6l-14.2-76.3-46.3-50.6s-10.3 48.1-10.5 49.1-.7 2.5-1.3 3.9c-.6 1.3-47.6 81.1-47.6 81.1-6.2 10.4-19.7 13.9-30.2 7.7-10.4-6.2-13.9-19.7-7.7-30.2l41.8-70.3 33.6-152.1-22.6 18-12.5 55.1c0 7.8-8.8 12.9-16.6 11.2-7.8-1.7-12.4-4.9-10.6-17.1h.2s13-61.2 14-63.9c.9-2.7 2.3-5.7 3.7-6.8 9.4-7.6 45.8-37.4 56.5-46 12.3-10 17.7-13.7 28.2-13.7 7.9 0 16.5 2.4 25.6 11.8 4 4.1 6.9 10.8 8.9 15 2 4.1 27.1 53.7 27.1 53.7l43.3 30c6.6 4.5 8.3 13.5 3.8 20.1z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-EHOje sc-bZQynM bskRhG"
    >
      <div
        class="sc-VigVT ikvPWi"
      >
        <div
          class="sc-jTzLTM dEKRsC"
        >
          <div
            class="sc-jzJRlG kkcrZI"
          >
            3520 NE Broadway, Portland, OR, USA 97232
          </div>
        </div>
        <div
          class="sc-kGXeez gskIpl"
        >
          <button
            aria-pressed="false"
            class="sc-bdVaJa sc-ckVGcZ kGRRRi"
          >
            Walk  0.3 miles to NE Sandy & Cesar Chavez Blvd
          </button>
          <button
            aria-pressed="false"
            class="sc-bdVaJa sc-kpOJdX livvEA"
          >
            <span
              class="sc-eHgmQL bQBtas"
            >
              Click to expand steps for this leg.
            </span>
            6 min walk
            <svg
              class="sc-dxgOiQ czQwWJ"
              expanded="false"
              fill="#484848"
              height="10"
              mode="WALK"
              viewBox="0 0 24 24"
              width="10"
            >
              <path
                d="M0 7.33L2.829 4.5l9.175 9.339L21.171 4.5 24 7.33 12.004 19.5z"
              />
            </svg>
          </button>
          <div
            class="sc-jKJlTe bBOoMC"
          />
        </div>
      </div>
    </div>
    <div
      class="sc-EHOje sc-gzVnrw fAkadq"
    >
      <button
        class="sc-bdVaJa sc-chPdSV pNQnJ"
      >
        <svg
          class="sc-kgoBCf kVSTZj"
          fill="#fff"
          height="15"
          role="img"
          viewBox="0 0 390 390"
          width="15"
        >
          <title>
            Frame this Itinerary Leg
          </title>
          <path
            d="M260.4 54.2L129.6 0 6.6 50.9v335.8l123-50.9L260.4 390l123-50.9V3.3l-123 50.9zM139.2 25l111.5 46.2v294L139.2 319V25zM25.9 63.9l94.1-39v294l-94.1 39v-294zm338.2 262.2l-94.1 39V71.2l94.1-39v293.9z"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
